### PR TITLE
New Resource: azurerm_log_profile

### PR DESCRIPTION
**Pull request by liemnotliam**
New Log Profile resource to address issue #1175

### CODE REVIEW

File: azurerm/resource_arm_log_profile.go [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+func resourceArmLogProfileCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).logProfilesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	categories := expandLogProfileCategories(d)
+	locations := expandLogProfileLocations(d)
+	retentionPolicy, err := expandAzureRmLogProfileRetentionPolicy(d)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile.go]: Combine them in which way, could you clarify?

File: azurerm/resource_arm_log_profile.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+		LogProfileProperties: &logProfileProperties,
+	}
+
+	_, createErr := client.CreateOrUpdate(ctx, name, parameters)
+	if createErr != nil {
+		return fmt.Errorf("Error Creating/Updating Log Profile %q: %+v", name, createErr)
+	}
+
+	// Wait for Log Profile to become available
+	err = resource.Retry(300*time.Second, retryLogProfilesClientGet(name, meta))
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile.go]: *question* Is the `Retry` here required for this resource?

File: azurerm/resource_arm_log_profile.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		}
+		return fmt.Errorf("Error making Read request on Azure Log Profile %q: %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	if props := resp.LogProfileProperties; props != nil {
+		d.Set("storage_account_id", props.StorageAccountID)
+		d.Set("service_bus_rule_id", props.ServiceBusRuleID)
+
+		locations := make([]string, len(*props.Locations))
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile.go]: Although defined as required in Azure REST spec, we might still need to `nil` check `props.Locations` to prevent potential crash.

```go
if props.Locations != nil {
  locations := …
  …
  d.Set("locations", locations)
}
```

File: azurerm/resource_arm_log_profile.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+		}
+
+		return logProfileRetentionPolicy, nil
+	}
+
+	return nil, fmt.Errorf("[ERROR] Retention policy must be set")
+}
+
+func flattenAzureRmLogProfileRetentionPolicy(input *insights.RetentionPolicy) []interface{} {
+	result := make(map[string]interface{})
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile.go]: There are several potential `nil` pointer crash possibilities here: `input`, `input.Enabled` and `input.Days`.

File: azurerm/resource_arm_log_profile_test.go [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+
+func testAccAzureRMLogProfile_basic(rInt int, rString string, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+	name     = "acctest%d-rg"
+	location = "%s"
+}
+
+resource "azurerm_storage_account" "test" {
+	name                     = "%s"
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile_test.go]: Naming convention `acctestsa%s`.

File: azurerm/resource_arm_log_profile_test.go [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+	resource_group_name      = "${azurerm_resource_group.test.name}"
+	location                 = "${azurerm_resource_group.test.location}"
+	account_tier             = "Standard"
+	account_replication_type = "GRS"
+}
+	
+resource "azurerm_log_profile" "test" {
+	name = "basic"
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile_test.go]: Append random number to `name` to prevent name conflict when running tests in parallel.

File: azurerm/resource_arm_log_profile_test.go [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+
+func testAccAzureRMLogProfile_servicebus(rInt int, rString string, location string) string {
+	return fmt.Sprintf(`
+resource "azurerm_resource_group" "test" {
+	name     = "acctest%d-rg"
+	location = "%s"
+}
+	
+resource "azurerm_servicebus_namespace" "test" {
+  name                = "a%s"
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile_test.go]: Naming convention `acctestsbns-%s`

File: azurerm/resource_arm_log_profile_test.go [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+  location            = "${azurerm_resource_group.test.location}"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+  sku                 = "standard"
+}
+
+resource "azurerm_servicebus_namespace_authorization_rule" "test" {
+  name                = "%s"
```
Comment (JunyiYi) [inline azurerm/resource_arm_log_profile_test.go]: Naming convention `acctestsbrule-%s`

File: website/docs/r/log_profile.html.markdown [hunk 9]
```diff
@@ -0,0 +1,104 @@
+---
```
Comment (JunyiYi) [inline website/docs/r/log_profile.html.markdown]: You should add this page URL to [the ToC](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/website/azurerm.erb#L692).

File: website/docs/d/log_profile.html.markdown [hunk 10]
```diff
@@ -0,0 +1,50 @@
+---
```
Comment (JunyiYi) [inline website/docs/d/log_profile.html.markdown]: This file should be added to the [data source ToC](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/website/azurerm.erb#L28).

File: azurerm/resource_arm_log_profile.go (hunk 2 above)
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile.go]: Yup, the client returns a few 404s before eventually returning a 200.

File: azurerm/resource_arm_log_profile_test.go (hunk 6 above)
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile_test.go]: The acceptance tests are combined into one run now. Azure doesn't allow for more than one log profile in a subscription so these tests shouldn't run in parallel.

File: website/docs/r/log_profile.html.markdown (hunk 9 above)
Comment (liemnotliam) [reply website/docs/r/log_profile.html.markdown]: added

File: website/docs/d/log_profile.html.markdown (hunk 10 above)
Comment (liemnotliam) [reply website/docs/d/log_profile.html.markdown]: added

File: azurerm/resource_arm_log_profile.go [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"days": {
```
[the remark this answers is not in the document]
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile.go]: fixed

File: azurerm/data_source_log_profile.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+func dataSourceArmLogProfileRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).logProfilesClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	resp, err := client.Get(ctx, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
```
[the remark this answers is not in the document]
Comment (liemnotliam) [reply azurerm/data_source_log_profile.go]: fixed

File: azurerm/resource_arm_log_profile.go (hunk 2 above)
Comment (JunyiYi) [reply azurerm/resource_arm_log_profile.go]: Hmm, that might be some API issue. Do we have bugs created in their [repository](https://github.com/Azure/azure-rest-api-specs)?

---
In reply to: [212493661](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1792#discussion_r212493661) [](ancestors = 212493661)

**Comment by liemnotliam** [comment]
The retries are being done in several other places that wait for the resource to become available:
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/dc324190abc9c787ba8c664b7ac85fd889a18d22/azurerm/resource_arm_policy_definition.go#L144
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/dc324190abc9c787ba8c664b7ac85fd889a18d22/azurerm/resource_arm_policy_assignment.go#L107
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/dc324190abc9c787ba8c664b7ac85fd889a18d22/azurerm/resource_arm_role_assignment.go#L110
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/dc324190abc9c787ba8c664b7ac85fd889a18d22/azurerm/resource_arm_storage_container.go#L111
- https://github.com/terraform-providers/terraform-provider-azurerm/blob/dc324190abc9c787ba8c664b7ac85fd889a18d22/azurerm/resource_arm_container_service.go#L238

There's more examples, but the retry logic is quite common for dealing with the quirks of the Azure API. If that's the only issue, is this PR ok to be approved then @JunyiYi ?

**Comment by liemnotliam** [comment]
Thanks for the comments and feedback @JunyiYi ! Is there scope for this to be included in the v1.14.0 from the Future milestone?

File: azurerm/resource_arm_log_profile_test.go [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func TestAccAzureRMLogProfile(t *testing.T) {
+	// NOTE: this is a combined test rather than separate split out tests due to
+	// Azure only being happy about provisioning one per subscription at once
+	// (which our test suite can't easily workaround)
+	testCases := map[string]map[string]func(t *testing.T){
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_log_profile_test.go]: (as above) is this still the case with this resource?

File: azurerm/resource_arm_log_profile_test.go [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+		"%s"
+	]
+	
+	storage_account_id  = "${azurerm_storage_account.test.id}"
+
+	retention_policy {
+		enabled = true
+		days = 7
+	}
+}
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_log_profile_test.go]: _minor_ can we switch to using spaces instead of tabs here?

File: azurerm/resource_arm_log_profile_test.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+		"%s",
+	]
+	
+	service_bus_rule_id = "${azurerm_servicebus_namespace_authorization_rule.test.id}"
+	
+	retention_policy {
+		enabled = false
+		days = 0
+	}
+}
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_log_profile_test.go]: _minor_ can we switch to using spaces instead of tabs here?

File: azurerm/resource_arm_log_profile_test.go [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+		"uksouth",
+		"ukwest",
+		"westcentralus",
+		"westus2",
+		"koreacentral",
+		"koreasouth",
+		"francecentral",
+		"francesouth",
+		"australiacentral",
+		"australiacentral2",
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_log_profile_test.go]: whilst this list passes in Azure Public - this'll fail in Azure China/Germany/Government etc - can we use the `testLocation()` and `testAltLocation()` methods to submit in both of these instead?

File: website/docs/d/log_profile.html.markdown [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+ 
+* `storage_account_id` - The resource id of the storage account in which the Activity Log is stored.
+
+* `service_bus_rule_id` - The service bus (or event hub) rule ID of the service bus (or event hub) namespace in which the Activity Log is streamed to.
+
+* `locations` - List of regions for which Activity Log events are stored or streamed.
+
+* `categories` - List of categories of the logs.
+
+* `retention_policy`- A retention policy for how long Activity Logs are retained in the storage account.
```
Comment (tombuildsstuff) [inline website/docs/d/log_profile.html.markdown]: Can we update this to be: 
```
* `retention_policy` - a `retention_policy` block as documented below
```

File: website/docs/r/log_profile.html.markdown [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+  ]
+	
+  locations = [
+    "centralus",
+    "eastus",
+    "eastus2",
+    "westus",
+    "northcentralus",
+    "southcentralus",
+    "global",
```
Comment (tombuildsstuff) [inline website/docs/r/log_profile.html.markdown]: for the purposes of this example we should be able to simplify this to just `global`?

File: azurerm/data_source_log_profile.go [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"storage_account_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"service_bus_rule_id": {
```
Comment (tombuildsstuff) [inline azurerm/data_source_log_profile.go]: can we make this `servicebus_rule_id` to match the other resources/Azure branding?

File: azurerm/data_source_log_profile.go [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+			return fmt.Errorf("Error: Log Profile %q was not found", name)
+		}
+		return fmt.Errorf("Error reading Log Profile: %+v", err)
+	}
+
+	d.SetId(*resp.ID)
+
+	if props := resp.LogProfileProperties; props != nil {
+		d.Set("storage_account_id", props.StorageAccountID)
+		d.Set("service_bus_rule_id", props.ServiceBusRuleID)
```
Comment (tombuildsstuff) [inline azurerm/data_source_log_profile.go]: can we make this `servicebus_rule_id` to match the other resources/Azure branding?

File: azurerm/resource_arm_log_profile.go [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"storage_account_id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: azure.ValidateResourceIDOrEmpty,
+			},
+			"service_bus_rule_id": {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_log_profile.go]: can we make this `servicebus_rule_id` to match the other resources/Azure branding?

File: azurerm/data_source_log_profile_test.go [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestAccDataSourceAzureRMLogProfile(t *testing.T) {
+	// NOTE: this is a combined test rather than separate split out tests due to
+	// Azure only being happy about provisioning one per subscription at once
+	// (which our test suite can't easily workaround)
```
[the remark this answers is not in the document]
Comment (liemnotliam) [reply azurerm/data_source_log_profile_test.go]: Yes, there's currently a limit of 1 log profile per subscription

File: azurerm/resource_arm_log_profile.go [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: azure.ValidateResourceIDOrEmpty,
+			},
```
[the remark this answers is not in the document]
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile.go]: No conflict, both can be specified

File: azurerm/resource_arm_log_profile_test.go (hunk 13 above)
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile_test.go]: Yup, there's currently a limit of 1 log profile per subscription

File: azurerm/resource_arm_log_profile.go [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"days": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  0,
```
[the remark this answers is not in the document]
Comment (liemnotliam) [reply azurerm/resource_arm_log_profile.go]: Yup, if the `retention_policy.enabled` is set to true then a value of 0 will retain the events indefinitely. Otherwise events will be kept for the number of days specified.

If the `retention_policy.enabled` is set to false then the value doesn't really matter but the SDK still expects an int to be passed. The default was intended for when retention policy is not enabled.

Ok to remove the default, but would need to mark the days field as required, not optional. Thoughts? @tombuildsstuff

File: azurerm/data_source_monitor_log_profile_test.go [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+					resource.TestCheckResourceAttrSet(dataSourceName, "retention_policy.0.days"),
+				),
+			},
+		},
+	})
+}
+
+func testAccDataSourceAzureRMMonitorLogProfile_storageaccountConfig(rInt int, rString string, location string) string {
+	return fmt.Sprintf(`
+        resource "azurerm_resource_group" "test" {
```
Comment (katbyte) [inline azurerm/data_source_monitor_log_profile_test.go]: Very minor and not a blocker, but could we left align theres:

```
	return fmt.Sprintf(`
resource "azurerm_resource_group" "test" {
```

File: website/docs/r/monitor_log_profile.html.markdown [hunk 26]
```diff
@@ -0,0 +1,99 @@
+---
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_monitor_log_profile"
+sidebar_current: "docs-azurerm-resource-monitor-log-profile"
+description: |-
+  Creates a Log Profile.
```
Comment (katbyte) [inline website/docs/r/monitor_log_profile.html.markdown]: This should be `Manages a`

File: website/docs/r/monitor_log_profile.html.markdown [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+---
+
+# azurerm_monitor_log_profile
+
+Creates a [Log Profile](https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-overview-activity-logs#export-the-activity-log-with-a-log-profile). A Log Profile configures how Activity Logs are exported.
```
Comment (katbyte) [inline website/docs/r/monitor_log_profile.html.markdown]: As should this

**Comment by liemnotliam** [comment]
Hey @katbyte ,

I've made the changes based on your comments.

The 409 conflict error typically indicates that there's an existing log profile so it can't create another due to that 1 log profile per subscription limitation.

**Comment by katbyte** [comment]
@liemnotliam,

Cleaning out our subscription and trying again. However if this is the case having the datasource and resource tests separate is going to cause issues with our nightly build, they will either need to be combined or locked so they don't run at the same time.

**Comment by katbyte** [comment]
@liemnotliam,

I am still unable to get it to work on our subscription. I suspect there is a log profile kicking around somewhere. Are you aware of where to find this in the portal?

However running the tests on another subscription, the test is showing a non-empty plan:

```
testing.go:513: Step 0 error: After applying this step, the plan was not empty:
               
               DIFF:
               
               UPDATE: azurerm_monitor_log_profile.test
                 locations.1790160812: "" => "westeurope"
                 locations.574794058:  "westeurope" => ""

```

**Comment by liemnotliam** [comment]
@katbyte 

The portal doesn't really provide an easy way to view an existing log profile. You can query it via az cli with `az monitor log-profiles list`.

I've combined the tests and with the test results below.
```bash
~/go/src/github.com/terraform-providers/terraform-provider-azurerm on log-profile*
TESTARGS="-run TestAccAzureRMMonitorLogProfile -count=1" make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run TestAccAzureRMMonitorLogProfile -count=1 -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
?       github.com/terraform-providers/terraform-provider-azurerm       [no test files]
=== RUN   TestAccAzureRMMonitorLogProfile
=== RUN   TestAccAzureRMMonitorLogProfile/basic
=== RUN   TestAccAzureRMMonitorLogProfile/basic/basic
=== RUN   TestAccAzureRMMonitorLogProfile/basic/servicebus
=== RUN   TestAccAzureRMMonitorLogProfile/basic/complete
=== RUN   TestAccAzureRMMonitorLogProfile/basic/disappears
=== RUN   TestAccAzureRMMonitorLogProfile/datasource
=== RUN   TestAccAzureRMMonitorLogProfile/datasource/eventhub
=== RUN   TestAccAzureRMMonitorLogProfile/datasource/storageaccount
--- PASS: TestAccAzureRMMonitorLogProfile (1070.07s)
    --- PASS: TestAccAzureRMMonitorLogProfile/basic (807.06s)
        --- PASS: TestAccAzureRMMonitorLogProfile/basic/basic (98.02s)
        --- PASS: TestAccAzureRMMonitorLogProfile/basic/servicebus (442.57s)
        --- PASS: TestAccAzureRMMonitorLogProfile/basic/complete (174.21s)
        --- PASS: TestAccAzureRMMonitorLogProfile/basic/disappears (92.27s)
    --- PASS: TestAccAzureRMMonitorLogProfile/datasource (263.01s)
        --- PASS: TestAccAzureRMMonitorLogProfile/datasource/eventhub (165.44s)
        --- PASS: TestAccAzureRMMonitorLogProfile/datasource/storageaccount (97.56s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm       1070.106s
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/authentication        0.023s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure 0.029s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/kubernetes    0.010s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/response      0.037s [no tests to run]
?       github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/set   [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/suppress      0.032s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/validate      0.026s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils 0.038s [no tests to run]
?       github.com/terraform-providers/terraform-provider-azurerm/version       [no test files]
```
